docs/frontend/Modern-JS-features.md

```diff
@@ -0,0 +1,498 @@
+* Lambda-type, 'arrow functions' similar to delegates in C#, and lambdas in Python or Java
+* Native classes - syntactic sugar to write classes rather than prototype objects.
+* Modules - to separate out code into manageable chunks.
+* `const` and `let` variable declarators to declare at block, rather than global, level
+
+These features allow quicker writing of Javascript, and writing of more understandable and readable code.
+
+### Tooling
+
+Because Javascript has introduced new features at such a fast pace in recent years, the support for new features in both Node and browsers is behind the current specification.  This means that, before we use features, we have to convert them into other code that performs the same task, but that doesn't use the modern syntax where it wouldn't be supported.
+
+#### Webpack
+Webpack is a bundling system that allows us to take separate javascript files, and use them as modules to create one overarching javascript application for the front end of the system.
+
+Webpack is run by Gulp using 'build', 'bundle', or 'watch' tasks - the 'watch' task will only run Webpack when the Javascript files are changed
+
+_Javascript files in the `gulp` directory do not bundle, however Gulp does support module use anyway, because it isn't browser-based_
+
+#### Babel
+Babel is a language translator for Javascript that can be set up to translate new syntax into old syntax equivalents.  For example, Babel will take a Javascript class, and covert it into a function with prototype:
+
+```javascript
+class Animal {
+    constructor(name) {
+        this.name = name
+    }
+
+    greet() {
+        console.log(`Hi, ${this.name}`);
+    }
+}
+
+const myDog = new Animal('dog');
+
+myDog.greet(); //outputs 'Hi, dog'
+```
+
+Translated version might be something like this, depending on support requirements:
+
+```javascript
+function Animal(name) {
+    this.name = name;
+}
+
+Animal.prototype.greet = function() {
+    console.log('Hi, ' + this.name)
+}
+
+var myDog = new Animal('dog');
+
+myDog.greet();
+```
+
+You can see that the above has done the following steps:
+* Replaced the class with a function - a function called with `new` in Javascript is a constructor.
+* Replaced the greet function with a similar function that is a member of `Animal.prototype` - `prototype` is an object used to create new objects.
+* Removed the string interpolation in `greet` and replaced it with concatenated strings
+* Change the use of `const` to `var`.  Some older browsers don't support `const`.
+
+> **Side note**: The above shows why `this` in Javascript behaves differently than expected.  Although in a class, you expect `this` to refer to an instance of a class, the way that objects are created with constructors is only _emulating_ classes - the functions are actually separate functions, so the reference to `this` which in Javascript refers to the scope the function exists within is not a reference to the class.
```

Comment:
`this` might just be my failure, but I found this passage hard to understand

Reply:
I suppose basically classes don't exist, but code for them does - if that helps?  When you have a class, it's actually sugar for a function.